src/Symfony/Component/HttpClient/HttpClientTrait.php

```diff
@@ -45,7 +45,11 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
         $options = self::mergeDefaultOptions($options, $defaultOptions, $allowExtraOptions);
 
         if (isset($options['json'])) {
-            $options['body'] = self::jsonEncode($options['json']);
+            $encodedJson = self::jsonEncode($options['json']);
+            if (!empty($options['body']) && $options['body'] !== $encodedJson) {
```

Comment:
I would always throw an exception even of both options have the same contents.

Reply:
That was my first solution, but it turns out `MockHttpClient` runs `prepareRequest` multiple times. This happens in `request` and then again in the provided response factory.
Because the reason for this PR is to avoid unexpected behavior for developers I decided to allow having the same value as a solution

Reply:
> MockHttpClient runs prepareRequest multiple times

can we change something so that calling prepareRequest multiple times doesn't end up creating "json" and "body" entries? e.g. unsetting "json" once it has been processed?

Reply:
We can do this in the response factory, but that seems a strange place. If you think this is ok, I can add it there.
Unsetting json in `HttpClientTrait` breaks another test that checks a value in json in `MockResponse->getRequestOptions`

Reply:
please update both places and I'll review the patch :)

Reply:
can't we fix MockHttpClient to prepare the request only once ?

Reply:
prepareRequest needs to be idempotent, so yes we can but that wouldn't fix the root issue

Reply:
```diff
diff --git a/src/Symfony/Component/HttpClient/HttpClientTrait.php b/src/Symfony/Component/HttpClient/HttpClientTrait.php
index 60d80bdf63..ca5517ae83 100644
--- a/src/Symfony/Component/HttpClient/HttpClientTrait.php
+++ b/src/Symfony/Component/HttpClient/HttpClientTrait.php
@@ -45,11 +45,11 @@ trait HttpClientTrait
         $options = self::mergeDefaultOptions($options, $defaultOptions, $allowExtraOptions);
 
         if (isset($options['json'])) {
-            $encodedJson = self::jsonEncode($options['json']);
-            if (!empty($options['body']) && $options['body'] !== $encodedJson) {
-                throw new InvalidArgumentException('Define either the "json" or the "body" option, setting both different is not supported.');
+            if (!empty($options['body']) ) {
+                throw new InvalidArgumentException('Define either the "json" or the "body" option, setting both is not supported.');
             }
-            $options['body'] = $encodedJson;
+            $options['body'] = self::jsonEncode($options['json']);
+            unset($options['json']);
             $options['headers']['content-type'] = $options['headers']['content-type'] ?? ['application/json'];
         }
 
diff --git a/src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php b/src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php
index 30673a2568..3b49a49e48 100644
--- a/src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php
+++ b/src/Symfony/Component/HttpClient/Tests/HttpClientTraitTest.php
@@ -201,7 +201,7 @@ class HttpClientTraitTest extends TestCase
 
     /**
      * @expectedException \Symfony\Component\HttpClient\Exception\InvalidArgumentException
-     * @expectedExceptionMessage Define either the "json" or the "body" option, setting both different is not supported
+     * @expectedExceptionMessage Define either the "json" or the "body" option, setting both is not supported.
      */
     public function testSetJSONAndBodyOptions()
     {
```

This will break the next line though, because `getRequestOptions` no longer contains 'json':
https://github.com/symfony/symfony/blob/8da76862fa10001dfb021254af9b3eaba47792fb/src/Symfony/Component/HttpClient/Tests/ScopingHttpClientTest.php#L76

Reply:
Another way of dealing with this all might be to remove the 'special' option json and have a class for json data that can be set to body, which will be processed by `HttpClientTrait->normalizeBody`. For example:
```php

namespace Symfony\Component\HttpClient;

interface RequestBody {
    public function getBody();
}

class JsonRequestBody implements RequestBody {
/// ....
}
```
Then the option can be set 
```php
$option['body'] = new JsonRequestBody(['foo' => 'bar']);
```

This will also break the tests at https://github.com/symfony/symfony/blob/8da76862fa10001dfb021254af9b3eaba47792fb/src/Symfony/Component/HttpClient/Tests/ScopingHttpClientTest.php#L76 though.

Not having the magic options seems clearer, but requiring to use JsonRequestBody seems more difficult...

Reply:
I'm OK with your patch above, just this change: `if (null !== $options['body'] ?? null) {`
then please suggest changes to fix the tests

Reply:
The default for body is '' in `HttpClientInterface::OPTIONS_DEFAULTS`, that's why I checked for empty. Changing the default option to `null` there breaks a lot.
I'll change it to  
`if (isset($options['body']) && '' !== $options['body']) {`

Reply:
sure